controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +204,143 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !ok1 || !ok2 {
+			// All kcp cloned infra machines should have this annotation.
+			// Missing the annotation may be due to older version machines or adopted machines.
+			// Should not be considered as mismatch.
+			return true
+		}
+
+		// Check if the machine's infrastructure reference has been created from the current KCP infrastructure template.
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
+			clonedFromGroupKind != kcp.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			return false
+		}
+		return true
+	}
+}
+
+// MatchesKubernetesVersion returns a filter to find all machines that match a given Kubernetes version.
+func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		if machine.Spec.Version == nil {
+			return false
+		}
+		return *machine.Spec.Version == kubernetesVersion
+	}
+}
+
+// MatchesKubeadmBootstrapConfig checks if machine's KubeadmConfigSpec is equivalent with KCP's KubeadmConfigSpec.
+func MatchesKubeadmBootstrapConfig(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane, cluster clusterv1.Cluster) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		kcpConfigSpecLocal := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+
+		// Machine's init configuration is nil when machine is the control plane is already initialized.
+		if bootstrapObj.Spec.InitConfiguration == nil {
+			kcpConfigSpecLocal.InitConfiguration = nil
+		}
+
+		// Machine's join configuration is nil when a machine is the first machine in the control plane.
+		if bootstrapObj.Spec.JoinConfiguration == nil {
+			kcpConfigSpecLocal.JoinConfiguration = nil
+		} else if kcpConfigSpecLocal.JoinConfiguration == nil {
+			// If KCP join configuration is not present, set machine join configuration to nil (nothing can trigger rollout here).
+			bootstrapObj.Spec.JoinConfiguration = nil
+		}
+
+		if bootstrapObj.Spec.InitConfiguration != nil && kcpConfigSpecLocal.InitConfiguration != nil {
+			bootstrapObj.Spec.InitConfiguration.TypeMeta = kcpConfigSpecLocal.InitConfiguration.TypeMeta
+		}
+
+		if bootstrapObj.Spec.JoinConfiguration != nil && kcpConfigSpecLocal.JoinConfiguration != nil {
+			bootstrapObj.Spec.JoinConfiguration.TypeMeta = kcpConfigSpecLocal.JoinConfiguration.TypeMeta
+		}
```

Comment:
@vincepri Trying to find a good comment here.
In semantic merge there is this comment that explain setting it to nil: `The type meta for embedded types is currently ignored`
Any idea what it means?

Here, setting TypeMeta to nil to avoid comparing ApiVersion.

Reply:
How about something like `Clear up the TypeMeta information from the comparison. Today, the kubeadm types are embedded in CABPK and KCP types, and don't carry this information.`